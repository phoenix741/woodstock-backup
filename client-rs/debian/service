[Unit]
Description=Woodstock Client Daemon

[Service]
ExecStart=/usr/bin/ws_client_daemon

[Install]
WantedBy=network.target