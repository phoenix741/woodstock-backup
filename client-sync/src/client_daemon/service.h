#pragma once

#include <QtService>

#include "woodstockclient.h"

class ClientService : public QtService::Service
{
    Q_OBJECT
public:
    explicit ClientService(int &argc, char **argv);

protected:
    virtual ClientService::CommandResult onStart() override;
    virtual ClientService::CommandResult onStop(int &exitCode) override;

private:
    std::unique_ptr<WoodstockClient> m_client;
};
