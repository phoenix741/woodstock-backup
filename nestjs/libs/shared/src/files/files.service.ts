import { Injectable } from '@nestjs/common';
import { CoreBackupsService, CoreFilesService, JsFileManifest, JsFileManifestType } from '@woodstock/shared-rs';
import { Archiver } from 'archiver';
import { Stats } from 'fs';
import { AsyncIterableX, AsyncSink, create } from 'ix/asynciterable';
import { Readable } from 'stream';

@Injectable()
export class FilesService {
  constructor(
    private backupService: CoreBackupsService,
    private fileService: CoreFilesService,
  ) {}

  /**
   * List all file in the backup directory of the hostname that represent a share path
   * @param hostname The host name
   * @param backupNumber The backup number
   */
  listShares(hostname: string, backupNumber: number): Promise<string[]> {
    return this.backupService.getBackupSharePaths(hostname, backupNumber);
  }

  searchFiles(
    hostname: string,
    backupNumber: number,
    sharePath: string,
    path?: Buffer,
    recursive = false,
  ): AsyncIterableX<JsFileManifest> {
    const sink = new AsyncSink<JsFileManifest>();

    this.fileService.list(hostname, backupNumber, sharePath, path ?? Buffer.from(''), recursive, (err, manifest) => {
      if (manifest) {
        sink.write(manifest);
      } else {
        sink.end();
      }
    });

    return create(() => sink[Symbol.asyncIterator]());
  }

  /**
   * Create a stream that can be used to read the file describe by the manifest
   * @param manifest The file manifest
   */
  readFileStream(manifest: JsFileManifest): Readable {
    const sink = new AsyncSink<Buffer>();

    this.fileService.readManifest(manifest, (err, data) => {
      if (data) {
        sink.write(data);
      } else {
        sink.end();
      }
    });

    return Readable.from(create(() => sink[Symbol.asyncIterator]()));
  }

  async createArchive(archiver: Archiver, hostname: string, backupNumber: number, sharePath: string, path?: Buffer) {
    const manifests = this.searchFiles(hostname, backupNumber, sharePath, path, true);

    for await (const manifest of manifests) {
      const isRegular = manifest.stats?.type === JsFileManifestType.RegularFile;
      const isDirectory = manifest.stats?.type === JsFileManifestType.Directory;
      const isSymLink = manifest.stats?.type === JsFileManifestType.Symlink;

      if (isRegular) {
        archiver.append(this.readFileStream(manifest), {
          name: manifest.path.toString('utf-8'),
          date: manifest.stats?.lastModified ? new Date(manifest.stats.lastModified) : undefined,
          mode: manifest.stats?.mode,
          stats: {
            size: manifest.stats?.size ? Number(manifest.stats?.size) : 0,
            mode: manifest.stats?.mode,
            mtime: manifest.stats?.lastModified ? new Date(manifest.stats?.lastModified) : undefined,
            isFile: () => isRegular,
            isDirectory: () => isDirectory,
            isSymbolicLink: () => isSymLink,
            isBlockDevice: () => manifest.stats?.type === JsFileManifestType.BlockDevice,
            isCharacterDevice: () => manifest.stats?.type === JsFileManifestType.CharacterDevice,
            isFIFO: () => manifest.stats?.type === JsFileManifestType.Fifo,
            isSocket: () => manifest.stats?.type === JsFileManifestType.Socket,

            dev: manifest.stats?.dev ? Number(manifest.stats?.dev) : 0,
            ino: manifest.stats?.ino ? Number(manifest.stats?.ino) : 0,
            nlink: manifest.stats?.nlink ? Number(manifest.stats?.nlink) : 0,
            uid: manifest.stats?.ownerId,
            gid: manifest.stats?.groupId,
            rdev: manifest.stats?.rdev ? Number(manifest.stats?.rdev) : 0,
            atime: manifest.stats?.lastRead ? new Date(manifest.stats?.lastRead) : undefined,
            ctime: manifest.stats?.created ? new Date(manifest.stats?.created) : undefined,

            blksize: 0,
            blocks: 0,
            atimeMs: 0,
            mtimeMs: 0,
          } as Stats,
        });
      } else if (isSymLink && manifest.symlink) {
        archiver.symlink(manifest.path.toString('utf-8'), manifest.symlink?.toString('utf-8'), manifest.stats?.mode);
      }
    }
  }
}
