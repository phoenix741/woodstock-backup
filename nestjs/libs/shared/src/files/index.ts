export * from './files.module';
export * from './files.service';
