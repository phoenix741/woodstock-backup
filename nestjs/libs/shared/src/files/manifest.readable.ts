import { CoreFilesService, JsFileManifest } from '@woodstock/shared-rs';
import { Readable } from 'node:stream';

export class CallbackReadable extends Readable {
  private manifest: JsFileManifest;
  private fileService: any;
  private reading: boolean;

  constructor(manifest: JsFileManifest, fileService: CoreFilesService) {
    super();
    this.manifest = manifest;
    this.fileService = fileService;
    this.reading = false;
  }

  _read() {
    if (this.reading) return;
    this.reading = true;

    this.fileService.readFile(this.manifest, (err: Error, data: Buffer) => {
      if (err) {
        this.emit('error', err);
        return;
      }
      if (data) {
        this.push(data);
      }
      this.push(null); // Signal end of stream
    });
  }
}
