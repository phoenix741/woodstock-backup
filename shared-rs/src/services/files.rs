use std::{collections::HashMap, path::PathBuf};

use futures_util::{pin_mut, StreamExt};
use napi::{
  bindgen_prelude::Buffer,
  threadsafe_function::{
    ErrorStrategy::{self},
    ThreadsafeFunction, ThreadsafeFunctionCallMode,
  },
  Error, JsFunction, Result,
};
use tokio::io::AsyncReadExt;
use woodstock::{
  config::{Backups, Context, BUFFER_SIZE},
  utils::path::vec_to_path,
  FileManifest,
};

use crate::{config::context::JsBackupContext, models::JsFileManifest};

#[napi(js_name = "CoreFilesService")]
pub struct JsFilesService {
  backups: Backups,
  pool_path: PathBuf,
}

#[napi]
impl JsFilesService {
  #[napi(constructor)]
  #[must_use]
  pub fn new(context: &JsBackupContext) -> Self {
    let context: Context = context.into();

    Self {
      backups: Backups::new(&context),
      pool_path: context.config.path.pool_path.clone(),
    }
  }

  #[napi]
  pub fn list(
    &self,
    hostname: String,
    backup_number: u32,
    share_path: String,
    path: Buffer,
    recursive: bool,

    #[napi(ts_arg_type = "(err: null | Error, result: JsFileManifest | null) => void")]
    callback: JsFunction,
  ) -> Result<()> {
    let tsfn: ThreadsafeFunction<Option<JsFileManifest>, ErrorStrategy::CalleeHandled> =
      callback.create_threadsafe_function(0, |ctx| {
        let Some(value) = ctx.value else {
          return Ok(vec![]);
        };
        Ok(vec![value])
      })?;

    let backup_number = usize::try_from(backup_number)
      .map_err(|_| Error::from_reason("Backup number is too large".to_string()))?;
    let manifest = self
      .backups
      .get_manifest(&hostname, backup_number, &share_path);

    let path: Vec<u8> = path.into();
    let search_path = vec_to_path(&path);

    let tsfn = tsfn.clone();

    tokio::spawn(async move {
      let entries = manifest.read_manifest_entries();
      pin_mut!(entries);

      while let Some(entry) = entries.next().await {
        let path = entry.path();
        // If not recursif we keep only all files that is in the search path directory, else we keep all file that are in a lower level directory
        if (recursive && path.starts_with(&search_path)) || (!recursive && path == search_path) {
          let file = JsFileManifest::from(entry);

          tsfn.call(Ok(Some(file)), ThreadsafeFunctionCallMode::Blocking);
        }
      }

      tsfn.call(Ok(None), ThreadsafeFunctionCallMode::Blocking);
    });

    Ok(())
  }

  #[napi]
  pub fn read_manifest(
    &self,
    manifest: JsFileManifest,

    #[napi(ts_arg_type = "(err: null | Error, result: Buffer) => void")] callback: JsFunction,
  ) -> Result<()> {
    let tsfn: ThreadsafeFunction<Option<Vec<u8>>, ErrorStrategy::CalleeHandled> = callback
      .create_threadsafe_function(0, |ctx| {
        let Some(value) = ctx.value else {
          return Ok(vec![]);
        };
        Ok(vec![value])
      })?;

    let manifest: FileManifest = FileManifest {
      path: manifest.path.into(),
      hash: manifest.hash.into(),
      chunks: manifest.chunks.iter().map(|c| c.clone().into()).collect(),
      stats: None,
      symlink: Vec::new(),
      xattr: Vec::new(),
      acl: Vec::new(),
      metadata: HashMap::new(),
    };
    let pool_path = self.pool_path.clone();

    tokio::spawn(async move {
      let reader = manifest.open_from_pool(&pool_path);
      pin_mut!(reader);

      let mut buffer = vec![0; BUFFER_SIZE];

      loop {
        let Ok(size) = reader.read(&mut buffer).await else {
          tsfn.call(
            Err(Error::from_reason(
              "Error while reading the manifest".to_string(),
            )),
            ThreadsafeFunctionCallMode::Blocking,
          );
          break;
        };
        if size == 0 {
          break;
        }

        tsfn.call(
          Ok(Some(buffer[..size].to_vec())),
          ThreadsafeFunctionCallMode::Blocking,
        );
      }

      tsfn.call(Ok(None), ThreadsafeFunctionCallMode::Blocking);
    });

    Ok(())
  }
}
