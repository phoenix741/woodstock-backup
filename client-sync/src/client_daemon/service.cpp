#include "service.h"

ClientService::ClientService(int &argc, char **argv) : Service{argc, argv}, m_client(new WoodstockClient())
{
}

ClientService::CommandResult ClientService::onStart()
{
    qDebug() << "Service was started";

    m_client->listen();

    qDebug() << "Service is listening";
    return ClientService::CommandResult::Completed; //service is now assumed started
}

ClientService::CommandResult ClientService::onStop(int &exitCode)
{
    qDebug() << "Stop received";
    
    exitCode = 0;

    m_client->stop();

    qDebug() << "Stopped";
    return ClientService::CommandResult::Completed;
}
