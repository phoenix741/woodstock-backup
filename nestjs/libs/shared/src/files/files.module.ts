import { Module } from '@nestjs/common';
import { ApplicationConfigModule } from '../config';
import { ManifestModule } from '../manifest';
import { PoolModule } from '../pool';
import { ScannerModule } from '../scanner';
import { FilesService } from './files.service';

@Module({
  imports: [ApplicationConfigModule, ScannerModule, ManifestModule, PoolModule],
  providers: [FilesService],
  exports: [FilesService],
})
export class FilesModule {}
