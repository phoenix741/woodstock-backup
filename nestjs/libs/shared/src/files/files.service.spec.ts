import { Test, TestingModule } from '@nestjs/testing';
import archiver from 'archiver';
import { constants as constantsFs } from 'fs';
import { count, toArray } from 'ix/asynciterable';
import { FilesService } from './files.service';

import { CoreBackupsService, CoreFilesService, JsFileManifest } from '@woodstock/shared-rs';

describe('FilesService', () => {
  let service: FilesService;

  const mockBackupsService = {
    getBackupSharePaths: jest.fn(),
  };

  const mockFilesService = {
    list: jest.fn(),
    readManifest: jest.fn(),
  };

  const MANIFEST = [
    { path: Buffer.from('d0/ax'), stats: { mode: constantsFs.S_IFDIR } },
    { path: Buffer.from('d0'), stats: { mode: constantsFs.S_IFDIR } },
    { path: Buffer.from('d1/dd1/ddd1'), stats: { mode: constantsFs.S_IFDIR } },
    { path: Buffer.from('d1/dd1/ddd2'), stats: { mode: constantsFs.S_IFREG } },
    { path: Buffer.from('d1/dd1/ddd3'), stats: { mode: constantsFs.S_IFREG } },
    { path: Buffer.from('d1/dd2/ddd1'), stats: { mode: constantsFs.S_IFDIR } },
    { path: Buffer.from('d1/dd2/ddd1/abcd'), stats: { mode: constantsFs.S_IFREG } },
    { path: Buffer.from('d1/dd2/ddd1/efgh'), stats: { mode: constantsFs.S_IFREG } },
    {
      path: Buffer.from('d1/dd2/ddd1/ijkl'),
      symlink: Buffer.from('symlink'),
      stats: { mode: constantsFs.S_IFLNK },
    },
    { path: Buffer.from('d1/dd2/ddd1/mnop'), stats: { mode: constantsFs.S_IFREG } },
    { path: Buffer.from('d1/dd2/ddd2'), stats: { mode: constantsFs.S_IFDIR } },
    { path: Buffer.from('d1/dd2/ddd3'), stats: { mode: constantsFs.S_IFREG } },
    { path: Buffer.from('d2/dd1/ddd1'), stats: { mode: constantsFs.S_IFDIR } },
    { path: Buffer.from('d2/dd1/ddd1'), stats: { mode: constantsFs.S_IFREG } },
    { path: Buffer.from('d3/dd1/ddd1'), stats: { mode: constantsFs.S_IFDIR } },
  ] as JsFileManifest[];

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        FilesService,
        { provide: CoreBackupsService, useValue: mockBackupsService },
        { provide: CoreFilesService, useValue: mockFilesService },
      ],
    }).compile();

    service = module.get<FilesService>(FilesService);
  });

  describe('listShares', () => {
    it('list share from a directory', async () => {
      // GIVEN
      mockBackupsService.getBackupSharePaths = jest.fn(() => ['file1', 'file3']);

      // WHEN
      const res = await service.listShares('hostname', 1);

      // THEN
      expect(res).toMatchSnapshot('res');
      expect(mockBackupsService.getBackupSharePaths).toHaveBeenCalledWith('hostname', 1);
    });
  });

  describe('searchFiles', () => {
    it('search files in directory', async () => {
      // GIVEN
      mockFilesService.list = jest.fn((h, b, s, p, r, cb) => {
        MANIFEST.forEach((m) => cb(null, m));
        cb(null, null);
      });

      // WHEN
      const res = await toArray(service.searchFiles('hostname', 1, 'sharePath', Buffer.from('d1/dd2')));

      // THEN
      expect(res.map((v) => ({ ...v, path: v.path.toString('utf-8') }))).toMatchSnapshot('res');
      expect(mockFilesService.list).toHaveBeenCalledWith(
        'hostname',
        1,
        'sharePath',
        Buffer.from('d1/dd2'),
        false,
        expect.any(Function),
      );
    });
  });

  describe('readFileStream', () => {
    it('read file stream is special file', async () => {
      // GIVEN
      mockFilesService.readManifest = jest.fn((m, cb) => {
        cb(null, null);
      });

      // WHEN
      const res = service.readFileStream({
        path: Buffer.from('d1/dd2/ddd1/abcd'),
        stats: {
          mode: constantsFs.S_IFIFO,
          ownerId: 0,
          groupId: 0,
          size: 0n,
          compressedSize: 0n,
          lastModified: 0,
          lastRead: 0,
          created: 0,
          type: 0,
          rdev: 0n,
          dev: 0n,
          ino: 0n,
          nlink: 0n,
        },
        xattr: [],
        acl: [],
        chunks: [],
        symlink: Buffer.from(''),
        hash: Buffer.from(''),
        metadata: {},
      });

      // THEN
      expect(await count(res)).toBe(0);
    });

    it('read file stream regular file', async () => {
      // GIVEN
      mockFilesService.readManifest = jest.fn((m: JsFileManifest, cb) => {
        m.chunks.forEach((c) => cb(null, c));
        cb(null, null);
      });

      // WHEN
      const res = service.readFileStream({
        path: Buffer.from('d1/dd2/ddd1/abcd'),
        stats: {
          mode: constantsFs.S_IFREG,
          ownerId: 0,
          groupId: 0,
          size: 0n,
          compressedSize: 0n,
          lastModified: 0,
          lastRead: 0,
          created: 0,
          type: 0,
          rdev: 0n,
          dev: 0n,
          ino: 0n,
          nlink: 0n,
        },
        xattr: [],
        acl: [],
        chunks: [
          Buffer.from('chunks1'),
          Buffer.from('chunks2'),
          Buffer.from('chunks3'),
          Buffer.from('chunks4'),
          Buffer.from('chunks5'),
        ],
        symlink: Buffer.from(''),
        hash: Buffer.from(''),
        metadata: {},
      });

      // THEN
      for await (const chunk of res) {
        expect(chunk.toString()).toMatchSnapshot('chunk');
      }
    });
  });

  describe('createArchive', () => {
    let originalReadFileStream: FilesService['readFileStream'];

    beforeEach(() => {
      originalReadFileStream = service.readFileStream;
    });

    afterEach(() => {
      service.readFileStream = originalReadFileStream;
    });

    it('create archive', async () => {
      // GIVEN
      mockFilesService.list = jest.fn((h, b, s, p, r, cb) => {
        MANIFEST.forEach((m) => cb(null, m));
        cb(null, null);
      });
      const mockedArchiverInstance = {
        append: jest.fn(),
        symlink: jest.fn(),
      } as unknown as archiver.Archiver;
      service.readFileStream = jest.fn().mockImplementation((m) => m.path.toString());

      // WHEN
      await service.createArchive(mockedArchiverInstance, 'hostname', 1, 'sharePath', Buffer.from('d1/dd2'));

      // THEN
      expect(mockedArchiverInstance.append).toMatchSnapshot('append');
      expect(mockedArchiverInstance.symlink).toMatchSnapshot('symlink');
      expect(service.readFileStream).toMatchSnapshot('readFileStream');
    });
  });
});
