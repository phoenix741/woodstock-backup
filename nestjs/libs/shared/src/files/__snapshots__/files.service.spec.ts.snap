// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilesService createArchive create archive: append 1`] = `
[MockFunction] {
  "calls": [
    [
      "d1/dd2/ddd1/abcd",
      {
        "date": undefined,
        "mode": 32768,
        "name": "d1/dd2/ddd1/abcd",
        "stats": {
          "atime": undefined,
          "ctime": undefined,
          "dev": undefined,
          "gid": undefined,
          "ino": undefined,
          "isDirectory": [Function],
          "isFile": [Function],
          "isSymbolicLink": [Function],
          "mode": 32768,
          "mtime": undefined,
          "nlink": undefined,
          "rdev": undefined,
          "size": undefined,
          "uid": undefined,
        },
      },
    ],
    [
      "d1/dd2/ddd1/efgh",
      {
        "date": undefined,
        "mode": 32768,
        "name": "d1/dd2/ddd1/efgh",
        "stats": {
          "atime": undefined,
          "ctime": undefined,
          "dev": undefined,
          "gid": undefined,
          "ino": undefined,
          "isDirectory": [Function],
          "isFile": [Function],
          "isSymbolicLink": [Function],
          "mode": 32768,
          "mtime": undefined,
          "nlink": undefined,
          "rdev": undefined,
          "size": undefined,
          "uid": undefined,
        },
      },
    ],
    [
      "d1/dd2/ddd1/mnop",
      {
        "date": undefined,
        "mode": 32768,
        "name": "d1/dd2/ddd1/mnop",
        "stats": {
          "atime": undefined,
          "ctime": undefined,
          "dev": undefined,
          "gid": undefined,
          "ino": undefined,
          "isDirectory": [Function],
          "isFile": [Function],
          "isSymbolicLink": [Function],
          "mode": 32768,
          "mtime": undefined,
          "nlink": undefined,
          "rdev": undefined,
          "size": undefined,
          "uid": undefined,
        },
      },
    ],
    [
      "d1/dd2/ddd3",
      {
        "date": undefined,
        "mode": 32768,
        "name": "d1/dd2/ddd3",
        "stats": {
          "atime": undefined,
          "ctime": undefined,
          "dev": undefined,
          "gid": undefined,
          "ino": undefined,
          "isDirectory": [Function],
          "isFile": [Function],
          "isSymbolicLink": [Function],
          "mode": 32768,
          "mtime": undefined,
          "nlink": undefined,
          "rdev": undefined,
          "size": undefined,
          "uid": undefined,
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`FilesService createArchive create archive: readFileStream 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "path": {
          "data": [
            100,
            49,
            47,
            100,
            100,
            50,
            47,
            100,
            100,
            100,
            49,
            47,
            97,
            98,
            99,
            100,
          ],
          "type": "Buffer",
        },
        "stats": {
          "mode": Long {
            "high": 0,
            "low": 32768,
            "unsigned": false,
          },
        },
      },
    ],
    [
      {
        "path": {
          "data": [
            100,
            49,
            47,
            100,
            100,
            50,
            47,
            100,
            100,
            100,
            49,
            47,
            101,
            102,
            103,
            104,
          ],
          "type": "Buffer",
        },
        "stats": {
          "mode": Long {
            "high": 0,
            "low": 32768,
            "unsigned": false,
          },
        },
      },
    ],
    [
      {
        "path": {
          "data": [
            100,
            49,
            47,
            100,
            100,
            50,
            47,
            100,
            100,
            100,
            49,
            47,
            109,
            110,
            111,
            112,
          ],
          "type": "Buffer",
        },
        "stats": {
          "mode": Long {
            "high": 0,
            "low": 32768,
            "unsigned": false,
          },
        },
      },
    ],
    [
      {
        "path": {
          "data": [
            100,
            49,
            47,
            100,
            100,
            50,
            47,
            100,
            100,
            100,
            51,
          ],
          "type": "Buffer",
        },
        "stats": {
          "mode": Long {
            "high": 0,
            "low": 32768,
            "unsigned": false,
          },
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": "d1/dd2/ddd1/abcd",
    },
    {
      "type": "return",
      "value": "d1/dd2/ddd1/efgh",
    },
    {
      "type": "return",
      "value": "d1/dd2/ddd1/mnop",
    },
    {
      "type": "return",
      "value": "d1/dd2/ddd3",
    },
  ],
}
`;

exports[`FilesService createArchive create archive: symlink 1`] = `
[MockFunction] {
  "calls": [
    [
      "d1/dd2/ddd1/ijkl",
      "symlink",
      40960,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`FilesService listShares list share from a directory: res 1`] = `
[
  "file1",
  "file3",
]
`;

exports[`FilesService readFileStream read file stream regular file: chunk 1`] = `"chunks1"`;

exports[`FilesService readFileStream read file stream regular file: chunk 2`] = `"chunks2"`;

exports[`FilesService readFileStream read file stream regular file: chunk 3`] = `"chunks3"`;

exports[`FilesService readFileStream read file stream regular file: chunk 4`] = `"chunks4"`;

exports[`FilesService readFileStream read file stream regular file: chunk 5`] = `"chunks5"`;

exports[`FilesService searchFiles search files in directory: res 1`] = `
[
  {
    "path": "d1/dd2/ddd1",
    "stats": {
      "mode": Long {
        "high": 0,
        "low": 16384,
        "unsigned": false,
      },
    },
  },
  {
    "path": "d1/dd2/ddd2",
    "stats": {
      "mode": Long {
        "high": 0,
        "low": 16384,
        "unsigned": false,
      },
    },
  },
  {
    "path": "d1/dd2/ddd3",
    "stats": {
      "mode": Long {
        "high": 0,
        "low": 32768,
        "unsigned": false,
      },
    },
  },
]
`;
